torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -1717,81 +1717,244 @@ const Expr* polyGCD(const Polynomial* poly) {
+  HashProvider& hasher = e->hasher();
+  if (hasher.hash(divisor) != hasher.hash(multiplier)) {
+    if (divisor->isConstant() && multiplier->isConstant()) {
+      // If both are scalar we may be able to find a common factor which becomes
+      // 'scalar' of the term.
+      if (immediateEquals(evaluateOp(new Mod(multiplier, divisor)), 0)) {
+        Expr* c = evaluateOp(new Div(multiplier, divisor));
+        scalar = evaluateOp(new Mul(c, scalar));
+      } else {
+        return std::make_tuple(nullptr, nullptr, nullptr, nullptr);
+      }
+    } else {
+      return std::make_tuple(nullptr, nullptr, nullptr, nullptr);
+    }
+  }
+
+  denom = IRSimplifier::simplify(other);
+
+  return std::make_tuple(scalar, denom, divisor, mod_divisor);
+}
+
+// Search the polynomial for Terms that can be merged in
+// (1) Round + Mod pattern: (x/y) * y + x % y => RoundOff(x,y) + Mod(x, y) => x
+// (2) Mod round + Mod pattern: (x/y % z)*y + x%y => ModRound(x, y, z) + Mod(x,
+// y) => x % (y*z)
 const Expr* simplifyRoundModPattern(const Polynomial* poly) {
-  std::set<const Term*> rounds;
-  std::set<const Term*> mods;
+  std::vector<const Term*> rounds;
+  std::vector<const Term*> mods;
+  std::vector<const Term*> mod_rounds;
   std::vector<const Term*> others;
 
-  // Split out the RoundOffs and Mod operations so we can inspect.
+  // Split out the Mod, ModRounds and RoundOffs operations so we can inspect.
   for (auto* c : poly->variables()) {
     if (c->variables().size() > 1) {
-      others.push_back(c);
+      std::tuple<const Expr*, const Expr*, const Expr*, const Expr*> dmm =
+          modRound(c);
+      if (!std::get<0>(dmm)) {
+        others.push_back(c);
+      } else {
+        mod_rounds.push_back(c);
+      }
       continue;
     }
 
     const Expr* e = c->variables()[0];
 
     if (e->expr_type() == IRNodeType::kRoundOff) {
-      rounds.insert(c);
+      rounds.push_back(c);
     } else if (e->expr_type() == IRNodeType::kMod) {
-      mods.insert(c);
+      std::tuple<const Expr*, const Expr*, const Expr*, const Expr*> dmm =
+          modRound(c);
+      if (!std::get<0>(dmm)) {
+        mods.push_back(c);
+      } else {
+        mod_rounds.push_back(c);
+      }
     } else {
       others.push_back(c);
     }
   }
 
-  // Can't continue without at least one RoundOff and one Mod.
-  if (rounds.empty() || mods.empty()) {
+  // Can't continue without at least one RoundOff/ModRound and one Mod.
+  if ((rounds.empty() && mod_rounds.empty()) || mods.empty()) {
     return nullptr;
   }
 
   HashProvider& hasher = poly->hasher();
   bool didAnything = false;
+  std::vector<const Term*> mods_merged;
+  bool repeat = true;
+  // Repeat merging terms till there are no Mods or the terms cannot be merged
+  // any further.
+  while (!mods.empty() && repeat) {
+    repeat = false;
+    for (std::vector<const Term*>::iterator mit = mods.end() - 1;
+         mit != mods.begin() - 1;
+         mit--) {
```

Comment:
Why do we iterate to `mods.begin() - 1` here?

Reply:
Why in reversed order? In one iteration, if the mod term can be merged with other terms, I'll immediately do merging and erase it from the vector without breaking the for loop. So iterate mod items from `mods.end()-1` to `mods.begin()`; the exit check would be `mit != mods.begin()-1` (i.e., `mit >= mods.begin()`).

Reply:
I understand why it is in reversed order, my question was why `mods.begin() - 1` instead of `mods.begin()` - I was confused by '-1', but now I got it, thanks!